Migrate "Advanced Parameters > Team > Employees > Add new employee" page

rokaszygmantas opened the pull request:
<!--
Thank you for contributing to the PrestaShop project! 

Please take the time to edit the "Answers" rows with the necessary information.

Check out our contribution guidelines to find out how to complete it:
https://devdocs.prestashop.com/1.7/contribute/contribution-guidelines/#pull-requests
 -->

| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | Migration of employee add/edit form.
| Type?         | improvement
| Category?     | BO
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | https://github.com/PrestaShop/PrestaShop/issues/10503
| How to test?  | The page is accessible via `admin-dev/index.php/configure/advanced/employees/1/edit`, keep in mind that `1` is the employee ID in the URL, which can be changed. The page should behave similarly to the legacy employee form.
<!-- Click the form's "Preview" button to make sure the table is functional in GitHub. Thank you! -->

:exclamation: Please rebuild the assets before testing :exclamation: 

Additional functionality to test, that's not visible on the first look:
* After signing in to addons, the _sign in_ button changes to sign out and your addons email should be displayed above it.
* [Bugfix] When editing a superadmin employee (other than yourself), the multishop tree should be hidden. Also, when changing the profile to superadmin from any other, it should hide the multishop tree (and vice versa).
* [Bugfix] When changing employee profile, the "Default page" dropdown is updated based on selected profile page access. 
* Addons "Sign in" button is visible only when editing own account AND when your profile has access to `AdminModules`
* When toggling the navigation in any page in BO, it should remember navigation position (collapsed/expanded) and it should stay after reloading the page. This is implemented separately in migrated and legacy pages, and should work in both. I mean when clicking this: https://prnt.sc/mrq8sa
* When "Memorize selected language" option is enabled, language that you select on translatable fields will remain all over BO. Setting can be found below employees list: https://prnt.sc/mrthku
* There are different validation messages displayed when changing the password, as shown in following screenshots: https://prnt.sc/ms76qb https://prnt.sc/ms76un https://prnt.sc/ms76xk https://prnt.sc/ms770g
When there is an error related to password validation, the save button will not submit the form.

Points labeled with [Bugfix] are about functionality, which was in legacy logic, but was not working there.
<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/prestashop/prestashop/12214)
<!-- Reviewable:end -->

rokaszygmantas committed:
WIP employee form

rokaszygmantas committed:
Change method order

rokaszygmantas committed:
Fix Employee domain namespace

rokaszygmantas committed:
WIP employee form fields

rokaszygmantas committed:
Load some of the employee fields values to the form

rokaszygmantas committed:
Fix spacing

rokaszygmantas committed:
Make language choice provider more reusable

rokaszygmantas committed:
Show language choices in employee form

rokaszygmantas committed:
Add employee default tab choices

rokaszygmantas committed:
Add active and permission profile fields in employee form

rokaszygmantas committed:
Add multistore shop tree to employee form

rokaszygmantas committed:
Check if multistore input should be displayed

rokaszygmantas committed:
Add form rest tags

rokaszygmantas committed:
Complete clickable avatar type

rokaszygmantas committed:
Move form_rest to a better place

rokaszygmantas committed:
Improve employee avatar display

rokaszygmantas committed:
Add employee form JS

rokaszygmantas committed:
Implement new connection to addons modal and make it actually connect

rokaszygmantas committed:
WIP connect to addons form type

rokaszygmantas committed:
Make addons data provider service public

rokaszygmantas committed:
Show addons account information and sign out button when logged in to addons

rokaszygmantas committed:
Implement addons logout

rokaszygmantas committed:
Structurize employee edit page\'s JS

rokaszygmantas committed:
Add comment

rokaszygmantas committed:
Change employee form based on access restrictions

rokaszygmantas committed:
Render change password type

rokaszygmantas committed:
Move "change password" button out of the inputs panel

rokaszygmantas committed:
Implement reset password JS actions

rokaszygmantas committed:
Empty password inputs when they are hidden

rokaszygmantas committed:
Migrate password generation logic

rokaszygmantas committed:
Do not show generated password feedback on repeated password field

rokaszygmantas committed:
Fix typo

rokaszygmantas commented:
hi @matks, there is a dead link in legacy code: 

https://github.com/PrestaShop/PrestaShop/blob/develop/controllers/admin/AdminEmployeesController.php#L266

it links to the PrestaShop forums, but I think there are some unrecognized parameters in query string, which make the link broken:
https://www.prestashop.com/forums/index.php?app=core&module=usercp

it's used here, when clicking on the avatar: https://prnt.sc/mia0y3

I can change it to something that works, wdyt?

matks commented:
@rokaszygmantas I think we should either remove it or redirect to something working. Let's ask product team:

@colinegin The avatar on this page is clickable (see Rokas comment above) but the link is broken. It targets the prestashop forum on a 404 page. I guess it used to be a dedicated Forum page (maybe a  signin/signup page ?) that is not available anymore.

Do we replace the link for something else (PS forum homepage ? PS marketing website homepage ?) or do we just remove it and make the avatar not clickable ?

rokaszygmantas committed:
Allow disabling whole choice tree form type easily

rokaszygmantas committed:
Disable multishop tree when editing a superadmin

rokaszygmantas committed:
Allow enabling/disabling choice tree component

rokaszygmantas committed:
Enable/disable shop choice tree based on profile selection

rokaszygmantas committed:
WIP reload tabs dropdown when employee profile is changed

rokaszygmantas committed:
Fetch accessible tabs for chosen profile

rokaszygmantas committed:
Implement default tab dropdown reload when changing profile

rokaszygmantas committed:
Add employee creation form and form default values

rokaszygmantas committed:
Employee creation data handling

rokaszygmantas committed:
Fix page title in edit employee form

rokaszygmantas committed:
Implement employee edit data handling

rokaszygmantas committed:
Change edit route in employee grid toload the new edit page

rokaszygmantas committed:
Implement password hashing

rokaszygmantas committed:
Fix change password field to not be required

rokaszygmantas committed:
Fix licenses

rokaszygmantas committed:
Implement validation from legacy employee object model and WIP validation from legacy controller

rokaszygmantas committed:
Implement superadmin edit form access checking

rokaszygmantas committed:
Control access to addons sign in button

rokaszygmantas commented:
Hi @colinegin,

I have one small improvement suggestion for this page:

If an employee is a SuperAdmin, he can always access all shops (in case of multishop) and that cannot be changed.
Anyhow, the shop association tree is still visible if you are editing a SuperAdmin, but it does not allow changing the shop association (I mean this: https://prnt.sc/mo59q4). 

My suggestion is to hide the shop association tree when super admin profile is selected, and show it when any other profile is selected.

Let me know what you think.

Also I'm looking forward to get an answer to what @matks wrote above (where he mentioned you) - regarding the broken link.

Thanks!

rokaszygmantas committed:
Implement shop association validation

rokaszygmantas committed:
Fix shop association control to not be disabled

rokaszygmantas committed:
Cleanup

rokaszygmantas committed:
Don't allow selecting super admin profile for another employee when logged in as non-superadmin

colinegin commented:
Hello !

Sorry for the delay, do not hesitate to add the "waiting for PM" label so that I don't miss your questions :)

@matks, i guess it used to bring you to your forum profile (or to a sign up if you don't have any account). I don't know why the link is broken, but I've made a test, when uploading a picture to my forum account, it is also uploaded in my employee account. 
@ttoine what do you think ?

@rokaszygmantas i'm not sure of what is the expected behaviour.
Option 1 : we should be able to remove access to one or several shops even for superAdmin profiles (maybe it makes sense to be a super admin for only one of the shops ?). In this case, there is a bug as I have a success message but the action is not saved.

Option 2 : we should not be able to remove any access to super admin profiles. In this case, I agree with you, it does not make any sense to show him the shop association tree.
What do you think ?

I'm also pinging @marionf to have her opinion on it ! :)

marionf commented:
I think it makes sense to be able to create a superadmin employee for only one or some shops, I guess it's a multistore issue

rokaszygmantas commented:
@colinegin, @marionf thanks for replies.
Out of from what I found in the code, it's not a bug that we cannot remove shop access from superadmins, it's how the functionality was meant to be working. 

There is a lot of code in the legacy core that implies that superadmin has access to all shops, so things would break if we change that.

A few examples:
All legacy lists rendering implies that superadmin has access to all shops:
https://github.com/PrestaShop/PrestaShop/blob/c98a9bbc38fbcb186e0a3339f7ae494fa2a06c32/classes/controller/AdminController.php#L3099

Other places that check if employee has access to a shop or shop group are using this code, which says that if the employee is a superadmin, he has access to any shop:
https://github.com/PrestaShop/PrestaShop/blob/69fed488666ab18865c0214d54104b518f66cbc4/classes/Employee.php#L493


I suggest to go with Option 2 (to hide the shop association tree for super admins). :slightly_smiling_face:

rokaszygmantas committed:
Check if employee email is not used when saving

rokaszygmantas committed:
Implement last order, customer message, customer notification sync on employee update

ttoine commented:
This still possible to upload a picture on a forum profile. However, I don't know how it was managed in the past. The forum also support Gravatar, and when people don't have a picture, it display their initials.

rokaszygmantas committed:
Prevent changing profile, status and shop association when editing own account

rokaszygmantas committed:
Fix employee password verification

rokaszygmantas committed:
Update employee password in cookie when it is updated

matks commented:
@colinegin @ttoine I think we can guess safely that it used to redirect the shop employee to some forum page for hm. I suggest we put the signin page, it's an easy choice and makes sense. Is it OK for you ?

ttoine commented:
ok for me

colinegin commented:
I agree with both of you @matks and @rokaszygmantas !

rokaszygmantas commented:
@matks so which link is that? this one: https://www.prestashop.com/forums/login/ ?

matks commented:
@rokaszygmantas it seems https://www.prestashop.com/forums/register/ and https://www.prestashop.com/forums/login/ do not redirect you to homepage when you are already logged in 😞 

@ttoine Can we redirect to https://www.prestashop.com/forums/ then ? It's the only valid option I see

ttoine commented:
ok for me. let's keep it as simple as possible to maintain.

matks commented:
@rokaszygmantas https://www.prestashop.com/forums/ it is then 😄

rokaszygmantas committed:
Update link on clickable avatar

rokaszygmantas committed:
Migrate toggle navigation action in symfony controllers

rokaszygmantas committed:
Use migrated toggle navigation action in legacy controllers

rokaszygmantas committed:
Migrate change employee form language action for legacy pages

rokaszygmantas committed:
Save selected language choice in migrated forms

rokaszygmantas committed:
Hide shop tree when selecting superadmin profile

rokaszygmantas committed:
Fix password validation in JS

rokaszygmantas committed:
Restore invalid password error message

rokaszygmantas committed:
Don't check if generated password was submitted

rokaszygmantas committed:
Fix error after rebase

rokaszygmantas committed:
Remove legacy AdminEmployeesController and related templates

rokaszygmantas committed:
Change url in navigation to point to migrated controller

rokaszygmantas committed:
Allow submitting the form when change password control is hidden

rokaszygmantas committed:
Fix empty shop association warning when multishop is disabled

rokaszygmantas committed:
Fix legacy link generation parameters

rokaszygmantas committed:
Fix translations

rokaszygmantas committed:
Fix cs

rokaszygmantas commented:
@matks finally ready to review :slightly_smiling_face:

matks commented on admin-dev/themes/default/js/admin-theme.js:
```diff
@@ -164,14 +164,11 @@ $(document).ready(function() {
 
         $.ajax({
-            url: "index.php",
+            url: $(this).data('toggle-url'),
```
I'm wondering if it is useful to pass the target URL as data in the clicked element 🤔 . This would be useful if the URL could be different for multiple navbar, however there will always be only 1 navbar in PS BO, isn't it ?

rokaszygmantas replied:
@matks 
IMO it's good to have everything related to the navbar in the same place, to see what it depends on.
Anyway, what could be the alternatives to this?

matks replied:
Alternative is hardcoded URL in a JS constant, not dynamic 😅

rokaszygmantas replied:
so maybe this current solution is better? :stuck_out_tongue:

matks replied:
👍

matks commented on admin-dev/themes/default/template/nav.tpl:
```diff
@@ -1,5 +1,5 @@
 <nav class="nav-bar d-none d-md-block" role="navigation" id="nav-sidebar">
-	<span class="menu-collapse">
+	<span class="menu-collapse" data-toggle-url="{$toggle_navigation_url}">
```
Same wondering as above: is it useful to have this variable

matks commented on admin-dev/themes/new-theme/js/components/addons-connector.js:
```diff
@@ -0,0 +1,78 @@
+
+const $ = window.$;
+
+export default class AddonsConnector {
```
@PierreRambaud this is a nice rework of @rokaszygmantas however I'm wondering if we should go even further and only perform calls to Addons from the backend ?

sarjon replied:
also it's missing docs.

rokaszygmantas replied:
addons calls are performed from backend already, this request goes to the `AddonsController:loginAction` which sends a request to addons server

matks commented on admin-dev/themes/new-theme/js/components/addons-connector.js:
```diff
@@ -0,0 +1,78 @@
+        this.$addonsLoginButton.show();
+        $('button.btn[type="submit"]', this.addonsConnectFormSelector).hide();
+      }
+    }).done((response) => {
```
Maybe we should handle `fail()` usecase with a dedicated error message ? What happens if we dont handle `fail` ?

sarjon replied:
i'd also like to use `then()` instead of `done()` as it's part of `Promise` API, wdyt?

rokaszygmantas replied:
if we don't handle `fail()` it will probably just fail silently with an error in console. `fail()` is not likely to happen though, only if the addons server would fail to respond, should we take care of such edge case?

If user enters invalid credentials, he will get a nice error message already.

PierreRambaud replied:
We must take care in case the server take too much time to return a response, or even if the server crash.
Adding a generic case:
```js
$(document).ajaxError(() => {
  // Add an alert / flash message
});
```

rokaszygmantas replied:
I've just checked - this ajax request is actually sent to prestashop's backend, which then sends another request to addons API and handles any errors that occurred, so if addons server fails we will get an error message already

matks replied:
It does not change :) we must still handle the usecase "the shop BO takes too much time to return a response or crashes and returns a HTTP 500" with a nice error message. Allowing the script to fail with an error in console is going to make all JS on the page errored, and might freeze the page.

rokaszygmantas replied:
fixed

matks commented on admin-dev/themes/new-theme/js/components/change-password-handler.js:
```diff
@@ -0,0 +1,119 @@
+   *
+   * @param {number} strength
+   */
+  _getPasswordStrengthFeedback(strength) {
```
👍

matks commented on classes/Link.php:
```diff
@@ -783,6 +783,21 @@ public function getAdminLink($controller, $withToken = true, $sfRouteParams = ar
 
+                break;
+
+            case 'AdminEmployees':
```
👍

matks commented on src/Adapter/Employee/EmployeeFormAccessChecker.php:
```diff
@@ -0,0 +1,109 @@
+            ));
+        }
+
+        return $employeeId === $this->contextEmployeeProvider->getId();
```
I'm puzzled by the name of the function and what it does behind 🤔 can you explain it, maybe in phpDoc ?

rokaszygmantas replied:
it checks if the employee's access to the form is restricted, as explained in the interface's phpdoc: https://github.com/PrestaShop/PrestaShop/blob/57d41fbc686fc5a66b113fabc84e4ac1ecbbc1a0/src/Core/Employee/Access/EmployeeFormAccessCheckerInterface.php#L35

@matks, do you think it needs a better explanation there? :thinking: 

The code simply checks if provided `$employeeId` is the same as context employee ID, as you have different restrictions when editing own employee account.

matks replied:
`as you have different restrictions when editing own employee account` that was the part I was missing: what is means to have a "restricted access". Can you put this part in the interface comment ?

matks commented on src/Adapter/Employee/FormLanguageChanger.php:
```diff
@@ -0,0 +1,59 @@
+    /**
+     * {@inheritdoc}
+     */
+    public function changeLanguage($languageIsoCode)
```
Can we make it clear in the function name that this will impact the user cookies ?

matks commented on src/Adapter/Employee/FormLanguageChanger.php:
```diff
@@ -0,0 +1,59 @@
+use PrestaShop\PrestaShop\Core\Employee\FormLanguageChangerInterface;
+
+/**
+ * Class FormLanguageChanger responsible for changing employee form language.
```
Can you rewrite it so the confusion is not possible between "the language of the Employee Form" and "the Language used in Forms for the Employee" 😅

matks commented on src/Adapter/Employee/NavigationHandler.php:
```diff
@@ -0,0 +1,58 @@
+/**
+ * Class NavigationHandler handles collapsing/expanding the navigation for context employee.
+ */
+final class NavigationHandler implements NavigationHandlerInterface
```
`NavigationHandler` is a bit vague 😅

matks commented on src/Adapter/Employee/NavigationHandler.php:
```diff
@@ -0,0 +1,58 @@
+    /**
+     * {@inheritdoc}
+     */
+    public function toggleNavigation($shouldCollapse)
```
Same here: in a classic application, we would never modify cookies in Core/Adapter code. We would perform this at Controller level because cookies are a web-only system and Core/Adapter code should be agnostic (it should be able to run for both CLI apps and web apps).

So I think it would be good to highlight the fact cookies are being modified here, else developers might not expect that
```suggestion
    public function togglesNavigationMenuInCookies($shouldCollapse)
```

matks commented on src/Core/Employee/Access/ProfileAccessChecker.php:
```diff
@@ -0,0 +1,69 @@
+    /**
+     * {@inheritdoc}
+     */
+    public function canAccessProfile($profileId)
```
`canAccess` or `canEdit` ?

rokaszygmantas replied:
IMO it's more of `canAccess` than `canEdit`, because if you cannot access the profile you will not be able to edit an employee who is from this profile. Also, we are not editing profiles in employee form, so it would be confusing to check if we can edit the profile when we don't intend to.

matks commented on src/Adapter/Profile/Employee/CommandHandler/AddEmployeeHandler.php:
```diff
@@ -0,0 +1,132 @@
+    public function handle(AddEmployeeCommand $command)
+    {
+        if (!$this->profileAccessChecker->canAccessProfile((int) $command->getProfileId())) {
+            throw new InvalidProfileException('The provided profile is invalid');
```
This message is not accurate. It might be valid, however it cannot access the edited profile. That's different 😉

matks commented on src/Adapter/Profile/Employee/CommandHandler/AddEmployeeHandler.php:
```diff
@@ -0,0 +1,132 @@
+     *
+     * @throws EmailAlreadyUsedException
+     */
+    private function assertEmailIsUsed($email)
```
```suggestion
    private function assertEmailIsNotAlreadyUsed($email)
```
or
```suggestion
    private function assertNotADuplicateEmail($email)
```

matks commented on src/Adapter/Profile/Employee/CommandHandler/EditEmployeeHandler.php:
```diff
@@ -0,0 +1,198 @@
+    public function handle(EditEmployeeCommand $command)
+    {
+        if (!$this->profileAccessChecker->canAccessProfile((int) $command->getProfileId())) {
+            throw new InvalidProfileException('The provided profile is invalid');
```
Same

matks commented on src/Adapter/Profile/Employee/CommandHandler/EditEmployeeHandler.php:
```diff
@@ -0,0 +1,198 @@
+     */
+    private function updatePasswordInCookie(Employee $employee)
+    {
+        if ($employee->id == $this->contextEmployeeProvider->getId()) {
```
Would it make sense to call this function if `$employee->id` is different from `$this->contextEmployeeProvider->getId()` ?

matks commented on src/Core/Domain/Profile/Employee/ValueObject/LastName.php:
```diff
@@ -0,0 +1,99 @@
+/**
+ * Stores employee's last name
+ */
+class LastName
```
We have `PrestaShop\PrestaShop\Core\Domain\Customer\ValueObject\LastName` that is quite similar. However I'm not sure whether or not we should merge these 2 🤔

sarjon replied:
I was about to remove both `LastName` and `FirstName` in Customer domain.

matks replied:
@sarjon in order to make them global or do you think they are not relevant anymore ?

matks commented on src/Core/Domain/Profile/Employee/ValueObject/FirstName.php:
```diff
@@ -0,0 +1,99 @@
+use PrestaShop\PrestaShop\Core\Domain\Profile\Employee\Exception\EmployeeConstraintException;
+
+/**
+ * Stores employee's first name
```
This class does not really store the first name. It carries the data, it does not store it.

matks commented on src/Core/Domain/Profile/Employee/ValueObject/FirstName.php:
```diff
@@ -0,0 +1,99 @@
+ * needs please refer to http://www.prestashop.com for more information.
+ *
+ * @author    PrestaShop SA <contact@prestashop.com>
+ * @copyright 2007-2018 PrestaShop SA
```
2019 😄

matks commented on src/Core/Domain/Profile/Employee/ValueObject/Password.php:
```diff
@@ -0,0 +1,83 @@
+    /**
+     * @var string maximum allowed password length
+     */
+    const MAX_LENGTH = 255;
```
@PierreRambaud Dont we have an issue with 255-long character strings and mysql encoding ? Should it impact this or not ?

matks commented on src/Core/Employee/Access/EmployeeFormAccessCheckerInterface.php:
```diff
@@ -0,0 +1,67 @@
+     *
+     * @return bool
+     */
+    public function isRestrictedAccess($employeeId);
```
I think it might be `hasRestrictedAccess` but still it's not clear to me what is a restricted access, so it will not be clear for another dev not prestashop-fluent 😃 can we find a better name ? Or explain it in phpDoc ?

rokaszygmantas replied:
couldn't find a better name for the method :thinking: but explained it in phpdoc

matks replied:
Solved in a previous comment: the ambiguous part is "what is a restricted access ? what does it mean to be restricted ?"

matks commented on src/Core/Employee/Access/EmployeeFormAccessCheckerInterface.php:
```diff
@@ -0,0 +1,67 @@
+ * needs please refer to http://www.prestashop.com for more information.
+ *
+ * @author    PrestaShop SA <contact@prestashop.com>
+ * @copyright 2007-2018 PrestaShop SA
```
2019
and check it everywhere please 😛

rokaszygmantas replied:
@matks I'm thinking to run the license fixing command `prestashop:licenses:update` but scared that it will update licenses that are out of scope of my PR, wdyt?

matks replied:
What about using a shell script to 1) extract all modified files by your PR in a whitelist 2) run `prestashop:licenses:update` 3) perform `git checkout` on all files NOT in your whitelist

So it will result in `prestashop:licenses:update` being applied only to your PR.

matks commented on src/Core/Employee/DefaultAvatarProviderInterface.php:
```diff
@@ -0,0 +1,40 @@
+/**
+ * Interface DefaultAvatarProviderInterface describes employee default avatar provider.
+ */
+interface DefaultAvatarProviderInterface
```
A Default Avatar Provider is very specific, do you think we could turn it into `AvatarProviderInterface` instead ? I'm a bit worried we have added 2 classes for a not-so-useful usecase 🤔 

With a `AvatarProviderInterface` it will be able to welcome methods such as `getAvatarUrl($id)` or `editAvatar(...)` later

matks commented on src/Core/Employee/EmployeeDataProviderInterface.php:
```diff
@@ -0,0 +1,42 @@
+     *
+     * @return string
+     */
+    public function getHashedPasswordById($employeeId);
```
```suggestion
    public function getHashedPasswordByEmployeeId($employeeId);
```
or
```suggestion
    public function getEmployeeHashedPassword($employeeId);
```

matks commented on src/Core/Employee/FormLanguageChangerInterface.php:
```diff
@@ -0,0 +1,40 @@
+    /**
+     * Change employee form language to given one.
+     *
+     * @param string $languageIsoCode two letter iso code of the language
```
2 or 3 🤔 ? I dont know well the ISO format but I think there is something about 2-letters code, 3-letters code, whether it's equivalent or 2 different formats for the same ...

rokaszygmantas replied:
there are different ISO formats: 
ISO 639-1 covers two letter codes: https://en.wikipedia.org/wiki/ISO_639-1
ISO 639-2 covers three letter codes: https://en.wikipedia.org/wiki/ISO_639-2

both can be called as ISO codes IMO :stuck_out_tongue:

matks commented on src/Core/Form/ChoiceProvider/LanguageByIdChoiceProvider.php:
```diff
@@ -35,16 +34,16 @@
+     * @param array $languages
      */
-    public function __construct(LanguageDataProvider $languageDataProvider)
+    public function __construct(array $languages)
```
I'm sorry but this will introduce a Breaking Change that we can avoid, so you cant do that.

You could create a new class and deprecate this one but I think here one easy compromise would be to extract $languages from $languageDataProvider so we dont keep $languageDataProvider as private var mais $languages instead

rokaszygmantas replied:
the idea was to make this choice provider not depend on `LanguageDataProvider`, but since it breaks BC I won't be able to do that :disappointed: I'll have to create another choice provider because I need _all_ languages choices while this choice provider only provides _active_ languages choices.

matks replied:
Unfortunately, yes, because `FormChoiceProviderInterface` does not allow to pass options. Maybe you can use a Decorator to make the 2 related and avoid duplicate code ?

Can you add a `@todo` statement in this new class, stating that is is very close to LanguageIdChoiceProvider and suggesting a refactor to merge these 2 ?

We could use a setter/configuratir to make the FormChoiceProvider stateful and obtain the desired behavior (something like `allowOnlyActiveLanguageChoices()` but that would seem wrong 🤔 ?

rokaszygmantas replied:
> Unfortunately, yes, because `FormChoiceProviderInterface` does not allow to pass options. Maybe you can use a Decorator to make the 2 related and avoid duplicate code ?

IMO decorating `LanguageByIdChoiceProvider` would be redundant in this case, as the code in decorator would be nearly similar to code in this new `LanguageChoiceProvider`. I'd need to inject all languages to the decorator, then check if the language exists in the choices array already and only if it doesn't exist -insert it there. Or ignore the code from the class that's being decorated and just use all language choices, which makes more sense, but neglects the idea behind decorator pattern.




> Can you add a `@todo` statement in this new class, stating that is is very close to LanguageIdChoiceProvider and suggesting a refactor to merge these 2 ?

Added this `@todo`.
 
> We could use a setter/configuratir to make the FormChoiceProvider stateful and obtain the desired behavior (something like `allowOnlyActiveLanguageChoices()` but that would seem wrong  ?

I agree, it seems wrong :confused: IMO even if the code is kind of duplicated, it's better to have two similar classes with an intention to merge them into one at some point, than to have workarounds in code.

matks commented on src/Core/Form/IdentifiableObject/DataHandler/EmployeeFormDataHandler.php:
```diff
@@ -0,0 +1,208 @@
+
+        if (!$this->hashing->checkHash($plainPassword, $oldPassword)) {
+            throw new EmployeeConstraintException(
+                'Old password is invalid.',
```
Not accurate. It's rather
```suggestion
                'Old and new passwords do not match.',
```

matks commented on src/PrestaShopBundle/Controller/Admin/Configure/AdvancedParameters/EmployeeController.php:
```diff
@@ -216,32 +223,194 @@ public function bulkDeleteAction(Request $request)
+     *
+     * @return Response
+     */
+    public function changeFormLanguageAction(Request $request)
```
I think we need some `@AdminSecurity` . Update rights, maybe ?

Employee John should not be able to modify other employee Jack without having the proper rights.

rokaszygmantas replied:
I don't think we need that, because employee form language is saved in local storage (cookies) and not directly related to any employee. Any logged in employee can overwrite that value. There is no employee account involved when changing the form language. Wdyt @matks?

matks replied:
The need for security is not related to the content of the controller action. Right now it modifies data into the cookie so indeed this information is not secured, but
- we might in the future modify this controller action to perform other operations such as database modification, event dispatching
- if a module hooks himself on this action it might do more things that require security

In fact, almost all Controller actions in the BO must be secured by default. Only some rare cases such as ajax GET actions which provide data accross all BO can be open.

rokaszygmantas replied:
I see, but which permissions should be checked when performing this action?

matks replied:
`@AdminSecurity("is_granted(['update'], request.get('_legacy_controller'))")` will do 👍

matks commented on src/PrestaShopBundle/Controller/Admin/Configure/AdvancedParameters/ProfilesController.php:
```diff
@@ -178,6 +179,24 @@ public function bulkDeleteAction(Request $request)
+     *
+     * @return JsonResponse
+     */
+    public function getAccessibleTabsAction(Request $request)
```
This route is used in Employee add/edit page ? We need to add security here.

Employee John should not be able to obtain other employee Jack tab list without proper rights. Maybe require READ rights for this controller ?

matks replied:
By proper rights, I mean: he has the ability to READ employee pages.

rokaszygmantas replied:
> This route is used in Employee add/edit page ?

yes, it's used to refresh accessible tabs dropdown when employee profile is changed in the form.



> By proper rights, I mean: he has the ability to READ employee pages.

@matks, If I check Employee READ rights here, does it make sense to keep this method in `ProfilesController`? IMO I should move this method to `EmployeeController` in this case.

matks replied:
> @matks, If I check Employee READ rights here, does it make sense to keep this method in `ProfilesController`? IMO I should move this method to `EmployeeController` in this case.

Good idea indeed 👍 
Can you check what happens if someone, probably by mistake, gives an employee the READ, CREATE and UPDATE rights for Profiles, but not for Employees ? I think the ajax route will return a 302. We just need to make sure it does not break the page behavior, instead the dropdown should be empty (bonus: add an error message but I dont want to spend too much time on an exotic usecase).

rokaszygmantas replied:
I've moved this action to `EmployeeController`. 

If you don't have READ rights for Employees now you will get a 500 response, but that's because of issue with legacy layout. You will not be able to access it if you can READ (or more) Profiles, but not Employees.

matks replied:
> If you don't have READ rights for Employees now you will get a 500 response, but that's because of issue with legacy layout

o_O we need to fix this (let's open an issue)

matks commented on src/PrestaShopBundle/Form/Admin/Configure/AdvancedParameters/Employee/EmployeeType.php:
```diff
@@ -0,0 +1,249 @@
+    {
+        $resolver
+            ->setDefaults([
+                // When is_restricted_access is set to true, the form will show fields differently:
```
👍

matks commented on src/PrestaShopBundle/Resources/config/services/adapter/employee.yml:
```diff
@@ -21,3 +21,50 @@ services:
+  prestashop.adapter.employee.form_language_changer:
+    class: 'PrestaShop\PrestaShop\Adapter\Employee\FormLanguageChanger'
+    arguments:
+      - '@prestashop.adapter.legacy.context'
```
Please finish the file with a blank line

rokaszygmantas replied:
there is a blank line actually, but github is not showing it. If you click "view file" it will show that this file contains 71 lines, while here it shows only 70.

matks replied:
Very confusing from github 😅

matks commented on src/PrestaShopBundle/Form/Admin/Type/TranslatableType.php:
```diff
@@ -44,12 +45,37 @@ class TranslatableType extends AbstractType
      */
-    public function __construct(array $locales)
-    {
+    public function __construct(
```
We can modify the constructor because this class has been introduced for 1.7.6 👍 https://github.com/PrestaShop/PrestaShop/pull/10704

matks commented on src/PrestaShopBundle/Form/Admin/Type/TranslatableType.php:
```diff
@@ -245,4 +277,24 @@ private function doesErrorFormAndCurrentFormMatches(FormInterface $errorForm, Fo
+     *
+     * @return array
+     */
+    private function getDefaultLocale(array $locales)
```
Will this change create a behavior change on pages that are already migrated and use it ?

rokaszygmantas replied:
@matks yes. This change brings this setting to life: https://prnt.sc/mulodb

If the mentioned setting is enabled, whenever you select a language in translatable input, it will memorize that language and will auto-select it in all translatable inputs all over back office.
It's a feature that was working on legacy pages but not implemented on migrated pages.

matks replied:
Nice ! 👍

matks commented on src/PrestaShopBundle/Resources/config/services/core/form/choice_provider.yml:
```diff
@@ -171,3 +176,13 @@ services:
+  prestashop.core.form.choice_provider.profile:
+    class: 'PrestaShop\PrestaShop\Core\Form\ChoiceProvider\ProfileChoiceProvider'
+    arguments:
+      - '@=service("prestashop.adapter.data_provider.profile").getProfiles(service("prestashop.adapter.legacy.context").getLanguage().id)'
```
Blank end line too

rokaszygmantas replied:
there is a blank line, same comment as above.

matks commented on src/PrestaShopBundle/Resources/views/Admin/TwigTemplateForm/bootstrap_4_layout.html.twig:
```diff
@@ -714,3 +724,68 @@
   </div>
 {% endblock generatable_text_widget %}
+
+{% block change_password_widget %}
```
I'm not sure we're going to reuse this widget 🤔

rokaszygmantas replied:
you never know :smile: 
however, is it bad to have it as a widget?

matks replied:
I'm thinking maybe we have a premature abstraction usecase here 😉.

I suggest we follow DRY and [WET](https://dev.to/wuz/stop-trying-to-be-so-dry-instead-write-everything-twice-wet-5g33) principles here and wait for a need to reuse it before creating a widget.

This `bootstrap_4_layout.html.twig` file contains highy-reusable and generic widgets such as `translatable_widget` so putting these quite specific widgets next to them feels wrong to me 🤔 

Moreover this file and its neighbors (`form_div_layout.html.twig` for example) could use some refactoring as they grew bigger and bigger overtime. So let's not make them even bigger if not needed 😛

matks commented on src/PrestaShopBundle/Resources/views/Admin/TwigTemplateForm/form_div_layout.html.twig:
```diff
@@ -229,6 +229,24 @@
   {{ block('button_widget') }}
 {%- endblock reset_widget -%}
 
+{%- block clickable_avatar_widget -%}
```
Same here: are we going to reuse it ?

rokaszygmantas replied:
answered same question above

matks commented on src/PrestaShopBundle/Resources/views/Admin/macros.html.twig:
```diff
@@ -219,8 +218,14 @@
     {% if extraVars.label is defined %}
+      {% set isRequired = form.vars.required|default(false) %}
+
+      {% if extraVars.required is defined %}
```
Can you explain the usecase where this is needed ?

rokaszygmantas replied:
"Change password" form type is shown as not required until you click on it, then it expands into a block with three required inputs. There was a problem with making those three inputs required while keeping the parent as not required, since in symfony forms if the parent form is not required, it's children will always be not required as well.

matks commented on tests-legacy/PrestaShopBundle/Controller/ControllerTest.php:
```diff
@@ -130,7 +130,6 @@ public function getControllersClasses()
             array('AdminStoresController'),
             array('AdminSuppliersController'),
             array('AdminAttributesGroupsController'),
-            array('AdminEmployeesController'),
```
Add the new test in https://github.com/PrestaShop/PrestaShop/blob/develop/tests-legacy/Integration/PrestaShopBundle/Controller/Admin/SurvivalTest.php

sarjon commented on admin-dev/themes/new-theme/.webpack/common.js:
```diff
@@ -35,7 +35,8 @@ module.exports = {
     contacts: './js/pages/contacts',
-    employee: './js/pages/employee',
+    employee: './js/pages/employee/index',
+    employee_form: './js/pages/employee/edit_form',
```
how about `form` instead of `edit_form`? so it's not coupled with "edit" action. @matks wdyt?

rokaszygmantas replied:
good idea, renamed it.

sarjon commented on admin-dev/themes/new-theme/js/components/addons-connector.js:
```diff
@@ -0,0 +1,78 @@
+   * Initialize events related to connection to addons.
+   */
+  initEvents() {
+    const t = this;
```
any reason not to use arrow function? :thinking:

sarjon commented on admin-dev/themes/new-theme/js/components/addons-connector.js:
```diff
@@ -0,0 +1,78 @@
+   *
+   * @private
+   */
+  _connect(addonsConnectUrl, formData) {
```
it's not private, since constructor does not limit accessibility.

sarjon commented on admin-dev/themes/new-theme/js/components/addons-connector.js:
```diff
@@ -0,0 +1,78 @@
+    }).done((response) => {
+      if (response.success === 1) {
+        location.reload();
+      } else {
```
seems that `else` can be easily removed.

rokaszygmantas replied:
`location.reload()` doesn't stop script execution so I'd rather keep the `else`

sarjon commented on admin-dev/themes/new-theme/js/components/change-password-handler.js:
```diff
@@ -0,0 +1,119 @@
+/**
+ * Generates a password and informs about it's strength.
+ */
+export default class ChangePasswordHandler {
```
as it is part of `js/components` it would be nice to see detailed documentation of how to reuse it.

rokaszygmantas replied:
I've expanded the js doc of all the components that were introduced in this PR

sarjon commented on admin-dev/themes/new-theme/js/components/change-password-handler.js:
```diff
@@ -0,0 +1,119 @@
+export default class ChangePasswordHandler {
+  constructor() {
+    // Minimum length of the generated password.
+    this.minLength = 8;
```
this could be made configurable maybe?

sarjon commented on admin-dev/themes/new-theme/js/components/change-password-handler.js:
```diff
@@ -0,0 +1,119 @@
+   * @param {number} passwordStrength
+   * @param {boolean} isPasswordValid
+   */
+  _displayFeedback($outputContainer, passwordStrength, isPasswordValid) {
```
same here, constructor does not limit accessibility to your methods.

sarjon commented on admin-dev/themes/new-theme/js/components/form/change-password-control.js:
```diff
@@ -0,0 +1,201 @@
+    this.oldPasswordInputSelector = '#employee_change_password_old_password';
+    this.newPasswordInputSelector = '#employee_change_password_new_password_first';
+    this.confirmNewPasswordInputSelector = '#employee_change_password_new_password_second';
+    this.generatedPasswordDisplaySelector = '#employee_change_password_generated_password';
```
First, i think this component shouldn't be part of `js/components` as it is coupled to Employee page.

Second, there are a lot of selectors, and as we discussed earlier, it would be better to extract them into `EmployeeMap` object so they can be resued for E2E tests.

rokaszygmantas replied:
decoupled the component from the page so it's more reusable now, also created a map for employee form selectors

sarjon commented on admin-dev/themes/new-theme/js/components/form/change-password-control.js:
```diff
@@ -0,0 +1,201 @@
+   *
+   * @private
+   */
+  _checkPasswordValidity() {
```
same here with private accessibility.

sarjon commented on admin-dev/themes/new-theme/js/components/form/change-password-control.js:
```diff
@@ -0,0 +1,201 @@
+  /**
+   * Hide an element.
+   */
+  _hide($el) {
```
missing docs blocks, same for `_show()`.

sarjon commented on admin-dev/themes/new-theme/js/components/password-validator.js:
```diff
@@ -0,0 +1,89 @@
+/**
+ * Class responsible for checking password's validity.
+ */
+export default class PasswordValidator {
```
this should be reusable, right? if so, docs are missing.

rokaszygmantas replied:
expanded docblocks in all JS components

sarjon commented on admin-dev/themes/new-theme/js/components/password-validator.js:
```diff
@@ -0,0 +1,89 @@
+    this.minPasswordLength = 8;
+
+    // Maximum allowed length for entered password
+    this.maxPasswordLength = 255;
```
would it make sense to make these 2 params configurable?

sarjon commented on admin-dev/themes/new-theme/js/pages/employee/EmployeeEditPage.js:
```diff
@@ -0,0 +1,147 @@
+/**
+ * Class responsible for javascript actions in employee edit page.
+ */
+export default class EmployeeEditPage {
```
is it only for edit?

sarjon commented on admin-dev/themes/new-theme/js/pages/employee/EmployeeEditPage.js:
```diff
@@ -0,0 +1,147 @@
+    this.shopChoiceTreeSelector = '#employee_shop_association';
+    this.shopChoiceTree = new ChoiceTree(this.shopChoiceTreeSelector);
+    this.employeeProfileSelector = '#employee_profile';
+    this.tabsDropdownSelector = '#employee_default_page';
```
same here, all selectors should be extracted into map object.

sarjon commented on admin-dev/themes/new-theme/js/pages/employee/EmployeeEditPage.js:
```diff
@@ -0,0 +1,147 @@
+    $(document).on('change', this.employeeProfileSelector, () => this._toggleShopTree());
+
+    // Reload tabs dropdown when employee profile is changed.
+    $(document).on('change', this.employeeProfileSelector, function () {
```
any reason not to use arrow function?

sarjon commented on src/Core/Domain/Profile/Employee/ValueObject/Email.php:
```diff
@@ -0,0 +1,131 @@
+/**
+ * Stores email address
+ */
+class Email
```
i think this VO should be in `Domain/ValueObject`, @matks wdyt?

sarjon commented on src/Core/Domain/Profile/Employee/ValueObject/LastName.php:
```diff
@@ -0,0 +1,99 @@
+<?php
+/**
+ * 2007-2018 PrestaShop
```
same here, old license.

sarjon commented on src/Core/Employee/Access/ProfileAccessCheckerInterface.php:
```diff
@@ -0,0 +1,42 @@
+     *
+     * @return bool
+     */
+    public function canAccessProfile($profileId);
```
it seems that `employeeId` would be useful here, it raises question: who do you check access against? you are forced to depend on context with this signature.

However, i'm not sure how easy it would be to implement that. @matks wdyt?

rokaszygmantas replied:
changed the method to accept employee id as an argument.

sarjon commented on src/Core/Form/ChoiceProvider/ProfileChoiceProvider.php:
```diff
@@ -0,0 +1,62 @@
+<?php
+/**
+ * 2007-2018 PrestaShop
```
double check all licenses.

sarjon commented on src/PrestaShopBundle/Resources/config/routing/admin/configure/advanced_parameters/employee.yml:
```diff
@@ -49,3 +50,36 @@ admin_employees_bulk_delete:
+    employeeId: \d+
+
+
+admin_employees_toggle_navigation:
```
too many lines above.

rokaszygmantas committed:
Small JS fixes

rokaszygmantas committed:
Rename JS file

rokaszygmantas committed:
Make addons connector methods private

rokaszygmantas committed:
Make min password length configurable

rokaszygmantas committed:
Fix JS according to feedback from @sarjon

rokaszygmantas committed:
Move JS selectors to a map, make components easier to reuse, document JS

rokaszygmantas committed:
Clarify comment

rokaszygmantas committed:
Small fixes after review part 1

rokaszygmantas committed:
Fixes after review part 2

rokaszygmantas committed:
Update licenses

rokaszygmantas committed:
Make default shop language auto selected in translatable inputs

rokaszygmantas committed:
Add employee pages to survival tests

rokaszygmantas committed:
Show a nice error message when ajax addons login fails

rokaszygmantas committed:
Fixes according to comments

rokaszygmantas committed:
Move "get accessible tabs" action to employee controller

rokaszygmantas committed:
Add security annotation for tabs retrieval action

matks commented on src/PrestaShopBundle/Resources/views/Admin/TwigTemplateForm/form_div_layout.html.twig:
```diff
@@ -229,6 +229,24 @@
+  </a>
+{%- endblock clickable_avatar_widget -%}
+
+{%- block addons_connect_widget -%}
```
Same here: if we dont reuse it, please do not make widgets as I'm worried about premature abstraction

matks commented on src/PrestaShopBundle/Resources/views/Admin/TwigTemplateForm/form_div_layout.html.twig:
```diff
@@ -412,3 +430,28 @@
   {%- endfor -%}
 {%- endblock attributes -%}
+
+{%- block addons_connected_information -%}
```
Same here: if we dont reuse it, please do not make widgets as I'm worried about premature abstraction

rokaszygmantas committed:
Update access rights for tabs retrieval action

rokaszygmantas committed:
Remove clickable avatar widget

rokaszygmantas committed:
Remove change password widget and clickable avatar form type

rokaszygmantas committed:
Remove addons connect widget

rokaszygmantas commented:
@matks I think I've covered all comments after review :slightly_smiling_face:

rokaszygmantas committed:
Fix cs

matks commented:
@rokaszygmantas tahnk you ! let's go to QA 😄

matks commented:
@LouiseBonnard This PR is validated, you can have a look at wording 😉

LouiseBonnard commented on src/PrestaShopBundle/Controller/Admin/Configure/AdvancedParameters/EmployeeController.php:
```diff
@@ -250,20 +443,64 @@ protected function getErrorForEmployeeException(EmployeeException $e)
+                EmployeeConstraintException::INVALID_FIRST_NAME => $this->trans(
+                    'The %s field is invalid.',
+                    'Admin.Notifications.Error',
+                    [sprintf('"%s"', $this->trans('Firstname', 'Admin.Global'))]
```
Shouldn't it be _First name_ instead?

rokaszygmantas replied:
Thanks - fixed. :+1:

LouiseBonnard commented on src/PrestaShopBundle/Controller/Admin/Configure/AdvancedParameters/EmployeeController.php:
```diff
@@ -250,20 +443,64 @@ protected function getErrorForEmployeeException(EmployeeException $e)
+                EmployeeConstraintException::INVALID_LAST_NAME => $this->trans(
+                    'The %s field is invalid.',
+                    'Admin.Notifications.Error',
+                    [sprintf('"%s"', $this->trans('Lastname', 'Admin.Global'))]
```
Same here, I'd rather write it _Last name_.

rokaszygmantas replied:
Thanks - fixed. :+1:

LouiseBonnard commented on src/PrestaShopBundle/Controller/Admin/Configure/AdvancedParameters/EmployeeController.php:
```diff
@@ -250,20 +443,64 @@ protected function getErrorForEmployeeException(EmployeeException $e)
+            'help_link' => $this->generateSidebarLink($request->attributes->get('_legacy_controller')),
+            'superAdminProfileId' => $configuration->get('_PS_ADMIN_PROFILE_'),
+            'getTabsUrl' => $this->generateUrl('admin_employees_get_tabs'),
+            'errorMessage' => $this->trans('You do not have permission to add this.', 'Admin.Notifications.Error'),
```
Please write "You need permission to..." instead - cf. https://github.com/PrestaShop/PrestaShop/issues/12235#issuecomment-456809001

rokaszygmantas replied:
Thanks - fixed. :+1:

LouiseBonnard commented on src/PrestaShopBundle/Form/Admin/Configure/AdvancedParameters/Employee/EmployeeType.php:
```diff
@@ -0,0 +1,265 @@
+                    $this->getNotBlankConstraint(),
+                    $this->getLengthConstraint(EmployeeEmail::MAX_LENGTH),
+                    new Email([
+                        'message' => $this->trans('This field is invalid', [], 'Admin.Notifications.Error'),
```
Why is it considered as invalid? If it is because of the length, perhaps we should add an element to help the user know what to change, like _This field cannot be longer than %limit% characters_ in _Admin.Notifications.Error_.

rokaszygmantas replied:
This particular message _This field is invalid_ is shown when a value entered in email field is not a valid email address (for example when @ symbol is missing).
It's not related to length - in case of too long value it currently shows "This field cannot be longer than 255 characters".

LouiseBonnard replied:
Alright, thank you @rokaszygmantas, let's keep _This field is invalid_ then!

rokaszygmantas committed:
Fix wording

rokaszygmantas commented:
@matks I've rebased this PR, so your approval was dismissed (again) :)

marionf commented:
@rokaszygmantas 

1. Help button opens a new window like below

![capture d'écran_1205](https://user-images.githubusercontent.com/13449658/54811992-56fb0500-4c8a-11e9-9388-69caca2ba23d.png)

It should be like this instead

![capture d'écran_1206](https://user-images.githubusercontent.com/13449658/54812020-6712e480-4c8a-11e9-88d9-c15bbae5eb71.png)

2. When I forgot to complete a field, message isn't correctly placed

![capture d'écran_1207](https://user-images.githubusercontent.com/13449658/54812254-fa4c1a00-4c8a-11e9-9ea0-48c7aba45be3.png)

3. Number of characters for password field is missing

![capture d'écran_1208](https://user-images.githubusercontent.com/13449658/54812539-983fe480-4c8b-11e9-8d01-140761f4962e.png)

![capture d'écran_1209](https://user-images.githubusercontent.com/13449658/54812542-9b3ad500-4c8b-11e9-8190-3cd00dbeb0da.png)

> When toggling the navigation in any page in BO, it should remember navigation position (collapsed/expanded) and it should stay after reloading the page. This is implemented separately in migrated and legacy pages, and should work in both. I mean when clicking this: https://prnt.sc/mrq8sa

It's not working for legacy and symfony pages

4. Let's say I have a logistician profile and can't access employee page like below

![capture d'écran_1211](https://user-images.githubusercontent.com/13449658/54814565-506f8c00-4c90-11e9-8553-9f5699ad34fd.png)

When I am connected with my logistician profile and try to edit my profile, it doesn't work
I am redirected to the default page defined in my account. It was working in legacy.

![capture d'écran_1212](https://user-images.githubusercontent.com/13449658/54814681-83b21b00-4c90-11e9-806d-306dec05f2c4.png)

rokaszygmantas committed:
Add missing help message

rokaszygmantas committed:
Make generated password field disabled

rokaszygmantas committed:
Fix employee access when editing own profile

rokaszygmantas commented:
@marionf thanks for feedback! Here are my comments:

1. I could not reproduce it - the help box opens nicely for me. Are you sure you've rebuilt the assets / cleared browser cache?

2. The problem is that when the page is scrolled down and the input is hidden under the page header, the tooltip still appears next to the input. Same is happening in other migrated forms, e.g. when editing a language. 
@matks not sure if I should fix this or something should be fixed globally, or even if this behavior should stay as is.
Example from languages edit page, when I've left the name field empty and clicked save: https://prnt.sc/n2jto4

3. Fixed.
4. Fixed.

matks commented on src/PrestaShopBundle/Resources/views/Admin/macros.html.twig:
```diff
@@ -52,8 +52,12 @@
-{% macro label_with_help(label, help, class = '') %}
-<label class="form-control-label {{ class }}">
+{% macro label_with_help(label, help, class = '', for = '', isRequired = false) %}
+<label{% if for is not empty %} for="{{ for }}"{% endif %} class="form-control-label {{ class }}">
```
`if for is not` I'm surprised it works 😄... but it does !

matks commented:
@marionf @rokaszygmantas Indeed we should fix it in a dedicated issue and pull request as this is a global bug

rokaszygmantas commented:
@marionf regarding the toggling of navigation menu, it's working for me on both migrated and symfony pages, I tried on a private window and a different browser, and it still worked. Maybe your browser cache is preventing it from working?

How it should work: when you collapse (or expand) the navigation menu and reload the page, the navigation menu should maintain the same state (collapsed/expanded) as it had before reloading.

marionf commented:
Thank you @rokaszygmantas it's all good for me

Issue about tooltip is here: #13040

@TristanLDD We have tooltips and a message under each field for the same information.
Don't know which one we should keep ?

![capture d'écran_1225](https://user-images.githubusercontent.com/13449658/54915142-c83ef000-4ef6-11e9-9aa6-55f02dab2022.png)

I think you have design feedbacks too ?

I am waiting before to put QA approved on this one, as there will be some changes after the feedbacks of @TristanLDD

rokaszygmantas commented:
@marionf I think it was agreed to have only small messages under fields and no tooltips, and I've just introduced the duplicate tooltips, which looks like a bug :slightly_smiling_face: but waiting for confirmation from @TristanLDD

TristanLDD commented:
Hi @marionf, indeed there are no more tooltips in this case, only the small messages. 

The only other design feedback is about the buttons "Cancel" and "save": there are no icons in these buttons, only text.

marionf commented:
@TristanLDD 

You don't want to adjust the length of the fields and the modales ? 

![capture d'écran_1215](https://user-images.githubusercontent.com/13449658/54926001-bae22f80-4f0f-11e9-81c3-59da9ad92cb7.png)

![capture d'écran_1213](https://user-images.githubusercontent.com/13449658/54926011-bf0e4d00-4f0f-11e9-8fa5-d8a958a35c45.png)

![capture d'écran_1210](https://user-images.githubusercontent.com/13449658/54926028-c9304b80-4f0f-11e9-8738-846b4cfe2d49.png)

rokaszygmantas committed:
Small fixes

rokaszygmantas commented:
@marionf 
the tooltips are removed now (only small text is remaining), and icons are removed from buttons. :slightly_smiling_face:

rokaszygmantas committed:
Add some comments

TristanLDD commented:
@marionf 

Here are the mockups for the 2 modals : 

![Capture d’écran 2019-03-26 à 11 14 51](https://user-images.githubusercontent.com/27484094/54989283-7d85ac80-4fb8-11e9-824b-90c37225091a.png)


![Capture d’écran 2019-03-26 à 11 15 06](https://user-images.githubusercontent.com/27484094/54989282-7d85ac80-4fb8-11e9-975b-1610f2f25518.png)

About the lenght of the fields, I actually need more time to think about it because we need an overall rule and I have to anticipate exceptions, so let's leave it like that for now.

rokaszygmantas committed:
Fix addons logout modal style

rokaszygmantas committed:
Fix addons sign in modal style

rokaszygmantas commented:
@marionf the modals style is fixed now :slightly_smiling_face:

marionf commented:
All good, thank you @rokaszygmantas :+1:

matks commented:
Thank you @rokaszygmantas and congrats for this big work !
Thank you @marionf for the hard QA work !
